File: javascript/DataStructures/DoublyLinkedList.js
class Node {
    constructor(value) {
        this.data = value;
        this.next = null;
        this.prev = null;
    }
}

class LinkedList {
    constructor() {
        this.head = null;
        this.tail = null;
    }

    pushFront(value) {
        let newNode = new Node(value);
        if (this.head === null) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            newNode.next = this.head;
            this.head.prev = newNode;
            this.head = newNode;
        }
    }

    pushBack(value) {
        let newNode = new Node(value);
        if (this.tail === null) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            this.tail.next = newNode;
            newNode.prev = this.tail;
            this.tail = newNode;
        }
    }

    popFront() {
        if (this.head === null) {
            throw new Error("Empty list. Cannot perform popFront!")
        }
        if (this.head.next === null) {
            this.head = null;
            this.tail = null;
        } else {
            this.head = this.head.next;
            this.prev = null;
        }
    }

    popBack() {
        if (this.head === null) {
            throw new Error("Empty list. Cannot perform popBack!")
        }
        if (this.head.next === null) {
            this.head = null;
            this.tail = null;
        } else {
            this.tail = this.tail.prev;
            this.tail.next = null;
        }
    }

    display() {
        if (this.head === null) {
            throw new Error("Empty list. Cannot perform display!")
        }

        let curr = this.head;
        let displayList = ""
        while (curr) {
            displayList += `${curr.data}->`;
            curr = curr.next;
        }
        displayList += "NULL";
        console.log(displayList);
    }
}

// MAIN AREA //
const list = new LinkedList();
list.pushBack(1);
list.pushBack(3);
list.pushBack(4);
list.pushBack(5);
list.pushFront(0);

try {
    list.popFront();
} catch (err) {
    console.log(err);
}

try {
    list.popBack();
} catch (err) {
    console.log(err);
}

list.display();

File: go/DataStructures/DoublyLinkedList/DoublyLinkedList.go
package DoublyLinkedList

import (
	"errors"
	"fmt"
)

type Node struct {
    data int
    next *Node
    prev *Node
}

func NewNode(value int) *Node {
    return &Node{
        data: value,
        next: nil,
        prev: nil,
    }
}

type LinkedList struct {
    head *Node
    tail *Node
}

func NewLinkedList() *LinkedList {
    return &LinkedList{
        head: nil,
        tail: nil,
    }
}

func (list *LinkedList) PushFront(value int) {
    newNode := NewNode(value)

    if list.head == nil {
        list.head = newNode
        list.tail = newNode
    } else {
        newNode.next = list.head
        list.head.prev = newNode
        list.head = newNode
    }
}

func (list *LinkedList) PushBack(value int) {
    newNode := NewNode(value)

    if list.tail == nil {
        list.head = newNode
        list.tail = newNode
    } else {
        list.tail.next = newNode
        newNode.prev = list.tail
        list.tail = newNode
    }
}

func (list *LinkedList) PopFront() error {
    if list.IsEmpty() {
        return errors.New("Empty list. Cannot perform PopFront")
    }
    if list.head.next == nil {
        list.head = nil
        list.tail = nil
    } else {
        list.head = list.head.next
        list.head.prev = nil
    }

    return nil
}

func (list *LinkedList) PopBack() error {
    if list.IsEmpty() {
        return errors.New("Empty list. Cannot perform PopBack")
    }
    if list.head.next == nil {
        list.head = nil
        list.tail = nil
    } else {
        list.tail = list.tail.prev
        list.tail.next = nil
    }

    return nil
}

func (list *LinkedList) IsEmpty() bool {
    return list.head == nil
}

func (list *LinkedList) Display() {
    curr := list.head

    for curr != nil {
        fmt.Printf("%d->", curr.data)
        curr = curr.next
    }
    fmt.Println("NULL")
}
